src/foam/dao/index/TreeIndex.java

```diff
@@ -95,6 +98,78 @@ protected Object[] simplifyPredicate(Object state, Predicate predicate) {
         state = ( (TreeNode) state ).lte((TreeNode) state, expr.getArg2().f(expr), indexer_);
         return new Object[] {state, null};
       }
+
+      if ( predicate.getClass().equals(In.class) && expr.getArg1().toString().equals(indexer_.toString()) ) {
+        var inPredicate = (In) predicate;
+        var arg2 = inPredicate.getArg2();
+
+        if ( arg2 instanceof Constant ) {
+          //FIXME: constant can be a list.
+          /**
+           * We can directly find key from tree.
+           * eg: MLang.IN(FObject.ID, "foo")
+           */
+          state = ((TreeNode) state).get((TreeNode) state, arg2.f(inPredicate), indexer_);
+          return new Object[] {state, null};
+        } else if ( arg2 instanceof ArrayConstant ) {
+          Object[] key = ((ArrayConstant) arg2).getValue();
+
+          if ( key == null || key.length == 0 ) {
+            /**
+             * Return nothing.
+             * eg: MLang.IN(FObject.ID, new Object[] {})
+             */
+            return new Object[] {null, null};
+          } else if ( key.length == 1 ) {
+            // eg: MLang.IN(FObject.ID, new Object[] {"foo"})
+            state = ((TreeNode) state).get((TreeNode) state, key[0], indexer_);
+            return new Object[] {state, null};
+          } else {
+            var treeRoot = (TreeNode) state;
+            long treeLevel = (long) treeRoot.level;
+            long treeheight = treeLevel; // treelevel is roughly equal to tree height.
+            long totalEdge = 2 << treeheight; // roughly estimation of total path. 
+            long averageEdge =  treeheight / 2; // roughly estimation of average edge jump required to search a key.
+
+            double estimateCardinality = (double) totalEdge / (double) treeRoot.size;
+
+            System.out.println("BBBBB average edge: " + averageEdge + ", require jump: " + averageEdge*key.length + ", total edge: " + totalEdge + ", estimateCardinality: " + estimateCardinality + ", tree level: " + treeLevel + ", tree height: " + treeheight);
+
+            if ( averageEdge*key.length < totalEdge>>2 && estimateCardinality > 0.25 ) {
+              /**
+               * precondition: edge jump is less than total edge and high cardinality.
+               * why need high cardinality?
+               *  The below will copy match data to a new tree. 
+               *  If the tail index contains many data, this optimization is not worth to do.
```

Comment:
Why does the size of the tail indices matter? They should be left as-is and only create a new tree at this level.